### PR TITLE
Refactored engine.io socket

### PR DESCRIPTION
**Pull request by nshaaban-cPacket**

### CODE REVIEW

**Comment by 1c3t3a** [comment]
I just fixed some errors on the current refactoring branch please rebase **all** open PRs onto the current refactoring branch.

File: src/engineio/socket.rs [hunk 1]
```diff
@@ -79,80 +37,153 @@ pub struct EngineSocket {
...
+    }
+
+    pub fn build(&self) -> Result<Socket<PollingTransport>> {
+        let (handshake, url) = self._build()?;
+
+        // Make a polling transport with new sid
+        let transport = PollingTransport::new(url, self.tls_config.clone(), self.headers.clone());
+
+        // If we can't upgrade or upgrade fails use polling
+        Ok(Socket::new(transport, handshake))
```
Comment (1c3t3a) [inline src/engineio/socket.rs:99]: If build was a consuming method (taking parameter `self`), we could get rid of all those `clone` statements on the member variables.

**Review by 1c3t3a — APPROVED** [review]
Looks good!

File: src/engineio/socket.rs (hunk 1 above)
Comment (nshaaban-cPacket) [reply src/engineio/socket.rs]: I was only able to get rid of one of the clone statements (as the shared _build function cannot consume self) by making them consume self. Changed the build methods to consume self so we are free to make more improvements in the future.

Comment (nshaaban-cPacket) [reply src/engineio/socket.rs]: refactored a bit more was able to remove another clone

**Comment by nshaaban-cPacket** [comment]
Merge this one first, higher priority than the others.

**Pull request merged**